### ISSUE TITLE
Error 32, os.rename, Python, file being used by another process

### ISSUE THREAD

**Issue by PhilHiTech** [EventGhost/EventGhost#402]
**Describe the bug**
12:12:59   Error while saving file
12:12:59   Traceback (most recent call last) (WIP-2020.04.30-23.04.40):
12:12:59     File "C:\Program Files (x86)\EventGhost\eg\Classes\Document.py", line 534, in WriteFile
12:12:59       os.rename(tmpPath, filePath)
12:12:59   WindowsError: [Error 32] Impossibile accedere al file. Il file è utilizzato da un altro processo

**To Reproduce**
Each time at operating system startup, after user sign in, and a few minutes after Eventghost starts.

**Expected behavior**
There should be no errors in Eventghost console.

**Additional context**
The egtree file is not on the system disk, and basically, Eventghost can't save the egtree file, and then saves it with the xml extension as a temporary file, each time a new xml file with a random name.

**EventGhost version**
Software
Program Version: WIP-2020.04.30-23.04.40
Build Time: gio, 30 apr 2020 23:04:40
Python Version: Stackless Python 2.7.12 final 0
wxPython Version: 3.0.2.0

System
Operating System: Microsoft Windows 10 Professional, 64-bit (build 18363)

**Comment by kdschlosser** [comment]
This is not a bug. You have something that is accessing the save file and holding it open. and that something is what is not allowing EG to save to it. You cannot write to a file that is opened up elsewhere.

also you are not running a release candidate you are running a test version. I do not know where this test version came from. I need to know this in order to see if there is a potential problem with it. If the version you are using has not been compiled by myself it is going to be hard for me to figure out if there is a problem with it because it could be modified in a manner that is causing the issue...

You also made the comment about the egtree file not being on the system disk. you need to elaborate on this. are you loading a save file from a network share? if that ios the case then you need to map the drive and access it using a drive letter.

**Comment by PhilHiTech** [comment]
Yes, it's probably not a bug, but I can't figure out which is the program at startup, which prevents the egtree file from being saved.
In fact, I have already added the Eventghost directory and its executable in the antivirus exceptions.
However, the egtree file is in the My Documents folder, which is located in the data disk ( D:\Documenti ), which is located together with the SSD system disk C:\ inside my notebook.
Also, the version I'm using is the "official" test version downloaded from
https://ci.appveyor.com/project/EventGhost/eventghost/branch/master/artifacts

**Comment by kdschlosser** [comment]
OK that version should be fine and without issue.

you can use performance monitor to see what is accessing the file. Windows also does spit out some wing dingers as far as errors are concerned. The actual error message may not be what is really happening. have you tried loading a test tree?? in EG click on file -> new. then save that file somewhere else. Exit EG and then reopen it and see if the same problem exists. you will have to make a change in order to save the file.

what EG does to ensure that an error in the program does end up corrupting your save file is it will write the data to a temporary file first. once it has completed that process without any errors it will then copy the file and rename it. In order to do this the destination cannot be in use at all.

I had to deal with a similiar issue with the config file. what I ended up doing was building the data in memory and then writing the file instead of using the whole temp file and rename scenario. moving/renaming a file has a different set of rules that have to be followed vs writing to the file. 

Try the temp file and let me know if that works or not. I am going to poke about the code some and make some changes to how the save file gets written to see if we can work around the issue.

I also ant you to be aware of Windows 10 and it's limitations as far as where things can be written to by an application. you may have to change where you are saving the file to.

**Comment by PhilHiTech** [comment]
The interesting thing is that the error occurs only at the Win10 startup and the user sign-in, then, after a few minutes, Eventghost manages to save the egtree file, although sometimes, I'm forced to save it manually ("autostart-macro.egtree").
As you say, I think that just the error message has to do with Windows 10, because at user sign-in, the svchost.exe process starts at 100% of 1 CPU core, and in addition it starts the "quick process scan" of the antivirus.
Unfortunately, I don't think the antivirus's "Quick Process Scan" can be deactivated, at least I couldn't do it.
As for Windows 10, I've already disabled many services and programs, I don't know what else to do...

**Comment by kdschlosser** [comment]
what antivirus are you using?

**Comment by kdschlosser** [comment]
My next question is what is telling EG to save right after you have logged into the computer?

**Comment by PhilHiTech** [comment]
Avira Antivirus Pro

**Comment by PhilHiTech** [comment]
![immagine](https://user-images.githubusercontent.com/57483887/80925221-51795500-8d8e-11ea-92b1-bca04fe614ae.png)
Well, my egtree file does several things, even at startup.
Using several EventGhost plugins and command line programs, it automates sending and receiving messages between Win10 and my Android phone ( Autoremote plugin and BurntToast ), and it does several other things anyway.

**Comment by PhilHiTech** [comment]
https://support.avira.com/hc/en-us/articles/360002544397-Why-can-t-I-disable-the-Quick-scan-function-in-Avira-Antivirus-
![immagine](https://user-images.githubusercontent.com/57483887/80925394-a073ba00-8d8f-11ea-97a7-809f78086202.png)

**Comment by kdschlosser** [comment]
are you a member of our support forum? I would like for you to PM me a copy of your EG save file.

**Comment by PhilHiTech** [comment]
No, I'm not a member of your support forum.
I'm just a novice macro script hobbyist, for my own use and enjoyment.
Thanks anyway for the availability, but in its entirety it is a complex set of scripts based on examples and tutorials found in various websites, localized in Italian, which would be long and complicated to explain...

**Comment by PhilHiTech** [comment]
So I think I'll try renaming the egtree file and moving it to another folder, to see if that's enough.

**Comment by kdschlosser** [comment]
I don't need an explanation I am looking for something inside of one of the scripts that is causing EG to save. EG is not going to save all on it's own The save button needs to be pressed or there is a script that is telling EG to save. This is why I do not understand why you are having an issue only when a user logs in. This is telling me that there is something inside of EG that you have set up that will cause EG to save. If that is the case then perhaps putting that bit on a timer so it gets done 5 or 10 minutes after a user logs in may be the solution.

**Comment by PhilHiTech** [comment]
However, I moved the egtree file to another folder ( D:\I\EventGhost_macro ), and I also changed the name of that file:

11:05:34   Error while saving file
11:05:34   Traceback (most recent call last) (WIP-2020.05.02-07.32.21):
11:05:34     File "C:\Program Files (x86)\EventGhost\eg\Classes\Document.py", line 534, in WriteFile
11:05:34       os.rename(tmpPath, filePath)
11:05:34   WindowsError: [Error 32] Impossibile accedere al file. Il file è utilizzato da un altro processo
11:05:34   Saved AutoRemote Configuration
(xml temp file created)

11:07:04   Error while saving file
11:07:04   Traceback (most recent call last) (WIP-2020.05.02-07.32.21):
11:07:04     File "C:\Program Files (x86)\EventGhost\eg\Classes\Document.py", line 534, in WriteFile
11:07:04       os.rename(tmpPath, filePath)
11:07:04   WindowsError: [Error 32] Impossibile accedere al file. Il file è utilizzato da un altro processo
11:07:04   Saved AutoRemote Configuration
(xml temp file created)

avscan.exe (Avira Antivirus) in background
appxsvc and appxsvc Win10 service in background

Could the AutoRemote plugin have something to do with this error?

**Comment by PhilHiTech** [comment]
Analyzing the differences between the two xml temp files, I would say that the problem is right at the moment of win10 logon.
The fact is that I also use Eventghost to determine when the logon happens, the idle periods, to start actions only when the user is in front of the screen.

**Comment by PhilHiTech** [comment]
I also noticed that at least in my case, the Autoremote plugin configuration window is not displayed correctly.
![immagine](https://user-images.githubusercontent.com/57483887/80955540-c3d64d80-8dff-11ea-90fd-2ecc25eabf1d.png)

**Comment by kdschlosser** [comment]
I believe it is the AutoRemote plugin causing the problem.

The Config dialog for the auto remote plugin has not been updated to work with EG 0.5 that is why you are seeing it displayed incorrectly.

I would need to look at the plugin to see what is happening. I am not sure exactly why it is telling EG to save.

**Comment by PhilHiTech** [comment]
Maybe the problem isn't the antivirus or win10...
Today I tried to deactivate and then re-enable the network interface, to simulate a disconnection, and even without restarting Eventghost, the error reappeared when Autoremote tries to save its configuration, after the reconnection.

**Comment by kdschlosser** [comment]
It has to be how AutoRemote is saving the data. I took a look at the plugin. and to be honest it is kind of messy and hard to follow. It may be better for you to ask the developer of autoremote to fix that issue and possibly update the configuration dialogs to support EventGhost 0.5. they may also want to spend some time to make the plugin python 3.6+ compatible as well. that is the direction we are heading in the months to come.

**Comment by PhilHiTech** [comment]
I'm back... I finally solved it!
Basically, in my egtree, I had inserted the Autoremote registration request of my mobile phone, when it had already been started from the mobile phone itself through the Tasker app...
By removing this request from the egtree, the error never appeared again.

**Comment by DanEdens** [comment]
I will speak with Joan and we'll get Autoremote updated. Been meaning to do that.

**Comment by kdschlosser** [comment]
Fantastic that you found the issue. Good deal!!
